### PR TITLE
Openfire TLS 1.3 workaround

### PR DESCRIPTION
**Pull request by sonnyp**
@netmikey could you review and test this?

### CODE REVIEW

**Comment by sonnyp** [comment]
You can 

1. Use xmpp.js git with this branch in your project
2. Copy paste `packages/tls/lib/Socket.js` in your `node_modules` 
3. Checkout this branch, tweak and use `packages/client/example.js` 

Please make sure to test before and after this changeset.

**Comment by netmikey** [comment]
Ok so here's what I'm at:

- it looks like xmpp.js 0.12.1 actually *does* work with direct tls (hadn't checked that earlier)
- neither "original" 0.12.1 nor 0.12.1 with this branch's `tls/lib/Socket.js` work with starttls

**Comment by sonnyp** [comment]
So I finally reinstalled Openfire to reproduce this.

There must be somethong wrong with how you switch xmpp.js- here are my results with direct TLS against openfire	

xmpp.js 0.12.1: KO - 10 out of 10
xmpp.js 0.12.0: KO - 10 out of 10 
xmpp.js [main](6bd956d633864fdc2e48e2fc7e03e342cdcdca7d): KO - 10 out of 10 
xmpp.js openfire-workaround: OK  - 10 out of 10

Would be really helpful and have saved tremendous amount of time if you could try to be a bit more thorough with your testing.

I forgot about STARTLS so this PR only addresses direct TLS for now but I will add it. It's the same problem anyway.

Please retry.

**Comment by sonnyp** [comment]
[`8f4a0e5` (#914)](https://github.com/xmppjs/xmpp.js/pull/914/commits/8f4a0e5b19319907438a6675673fe17cdd120595) adds the fix for starttls - would be great if you could test this branch with both direct TLS and starttls with your setup to confirm it fixes it.

Please make sure to test before/after.

Let me know if I can help. 

:crossed_fingers:

**Comment by netmikey** [comment]
Yes, sorry I forgot to check with direct TLS, my bad :(

As a side note: I don't have an easy way to install 0.12.0 anymore on my project. The reason being: I use `@xmpp/client` which depends on all of the @xmpp packages transitively. Since those dependencies aren't declared as fixed dependencies, but with `^`, npm will chose the latest 0.12 patch version it knows, which is 0.12.1 now. I would have to declare each @xmpp package (with a fixed dependency to 0.12.0) to really get version 0.12.0 back. It would maybe make things more reproducible if all @xmpp modules would refer to their siblings using fixed versions within one release.

But 0.12.0 isn't all that interesting anymore anyway. So here are my results now, everything checked multiple times. Tested against Openfire (configured to only allow TLSv1.3 on the direct tls and the starttls ports, using a real certificate):

- xmpp.js 0.12.1, direct tls: **OK**
- xmpp.js 0.12.1, starttls: **KO**
- xmpp.js 0.12.1 with `starttls` and `tls` modules copied from `openfire-workaround`, direct tls: **OK**
- xmpp.js 0.12.1 with `starttls` and `tls` modules copied from `openfire-workaround`, starttls: **OK**

For reference, here's how I reset my project to npm's xmpp.js 0.12.1:

    npm cache clean --force && rm -rf node_modules package-lock.json && npm install && npm start

File: packages/tls/lib/Socket.js [hunk 1]
```diff
@@ -28,7 +29,17 @@ class Socket extends EventEmitter {
...
+        return this.emit("connect");
+      }
+
+      // Waiting before sending the stream header improves compatibility
+      // with Openfire TLSv1.3 implementation. For more info, see:
+      // https://github.com/xmppjs/xmpp.js/issues/889#issuecomment-902686879
+      // https://github.com/xmppjs/xmpp.js/pull/912
+      this.timeout = setTimeout(() => {
+        this.emit("connect");
+      }, 1);
```
[the remark this answers is not in the document]
Comment (guusdk) [reply packages/tls/lib/Socket.js:42]: This problem is fixed in Openfire 4.8.0. That is not the latest release, but people are generally very slow to upgrade their server.

**Pull request merged**